src/thor/astar_bss.cc

```diff
@@ -0,0 +1,617 @@
+  float factor = 1.0f;
+  if (25000.0f < dist && dist < 100000.0f) {
+    factor = std::min(3.0f, 100000.0f / dist);
+  }
+  /* TODO - need a reliable density factor near the destination (e.g. tile level?)
+  // Low density - increase expansion within distance.
+  // High density - decrease expansion within distance.
+  if (density < 8) {
+    float f = 1.0f + (8.0f - density) * 0.125f;
+    factor *= f;
+  } else if (density > 8) {
+    float f = 0.5f + (15.0f - density) * 0.0625;
+    factor *= f;
+  }*/
+  // TODO - just arterial for now...investigate whether to alter local as well
+  pedestrian_hierarchy_limits_[1].expansion_within_dist *= factor;
+  bicycle_hierarchy_limits_[1].expansion_within_dist *= factor;
+}
+
+// Expand from the node along the forward search path. Immediately expands
+// from the end node of any transition edge (so no transition edges are added
+// to the adjacency list or EdgeLabel list). Does not expand transition
+// edges if from_transition is false.
+void AStarBSSAlgorithm::ExpandForward(GraphReader& graphreader,
+                                      const GraphId& node,
+                                      const EdgeLabel& pred,
+                                      const uint32_t pred_idx,
+                                      const bool from_transition,
+                                      const bool from_bss,
+                                      const sif::TravelMode mode,
+                                      const valhalla::Location& destination,
+                                      std::pair<int32_t, float>& best_path) {
+  const auto& current_costing =
+      (mode == TravelMode::kPedestrian ? pedestrian_costing_ : bicycle_costing_);
+  const auto& current_heuristic =
+      (mode == TravelMode::kPedestrian ? pedestrian_astarheuristic_ : bicycle_astarheuristic_);
+  auto& current_hierarchy_limits =
+      (mode == TravelMode::kPedestrian ? pedestrian_hierarchy_limits_ : bicycle_hierarchy_limits_);
+
+  // Get the tile and the node info. Skip if tile is null (can happen
+  // with regional data sets) or if no access at the node.
+  const GraphTile* tile = graphreader.GetGraphTile(node);
+  if (tile == nullptr) {
+    return;
+  }
+  const NodeInfo* nodeinfo = tile->node(node);
+  if (!current_costing->Allowed(nodeinfo)) {
+    return;
+  }
+
+  // Expand from end node.
+  uint32_t shortcuts = 0;
+  uint32_t max_shortcut_length = static_cast<uint32_t>(pred.distance() * 0.5f);
+  GraphId edgeid(node.tileid(), node.level(), nodeinfo->edge_index());
+
+  EdgeStatusInfo* current_es =
+      (mode == TravelMode::kPedestrian ? pedestrian_edgestatus_ : bicycle_edgestatus_)
+          .GetPtr(edgeid, tile);
+  const DirectedEdge* directededge = tile->directededge(nodeinfo->edge_index());
+
+  for (uint32_t i = 0; i < nodeinfo->edge_count(); ++i, ++directededge, ++edgeid, ++current_es) {
+    // Skip shortcut edges until we have stopped expanding on the next level.
```

Comment:
hi, yes, you're absolutely right... I admit that this is a stupid copy-paste of the original Astar algorithm... The reason why I leave it there is that we'd like to test with (auto+parking)... since this algo is dedicated to BSS, I should remove this...

Reply:
so, I suppose I should stop using hierarchy in bss_astar? I'll remove all codes concerning hierarchy and shortcuts...

Reply:
You still need to transition to edges on different hierarchies but you should not use hierarchy limits in BSS. Hierarchy limits will start to exclude minor roads and paths as you get further from the origin/destination. This works for driving routes, but not bicycle and pedestrian.

Reply:
Plz have a look at my last commit and tell me if I got your idea right :) thx https://github.com/valhalla/valhalla/pull/2031/commits/2bd65252c748dec0747e965414e88a32b12ad34d